src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,249 @@
+  }
+}
+
+export function shouldPropertyAutocomplete(input: string) {
+  if (!isAccessingObjProperty(input)) {
+    return false;
+  }
+
+  const parentExpression = getParentExpression(input);
+
+  if (!parentExpression) {
+    return false;
+  }
+
+  // Right now, this only works for direct property access (e.g., a.b)
+  // and doesn't support chained expressions (e.g., a.b.c.d). This is because
+  // we don't (yet) have the chained values in the fetched frame scope data.
+  return isIdentifier(parentExpression);
+}
+function shouldVariableAutocomplete(input: string) {
+  return isIdentifier(input);
+}
+
+function getBinding(name: string, frameScope: FrameScope) {
+  return frameScope.scope.bindings.find(b => b.name === name);
+}
+function getBindingNames(scope: FrameScope): string[] {
+  if (!scope?.scope) {
+    return [];
+  }
+
+  return scope.scope.bindings.map(b => b.name);
+}
+function getGlobalVariables(scopes: Scope) {
+  const variableNames = [];
+  const globalObject = scopes.parent.object;
+
+  if (globalObject) {
+    const properties = getPropertiesForObject(globalObject._object);
+    variableNames.push(...properties);
+  }
+
+  return variableNames;
+}
+
+export function getLastToken(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(input.lastIndexOf("[") + 1);
+  } else {
+    return input.slice(input.lastIndexOf(".") + 1);
+  }
+}
+
+export function getCursorIndex(value: string) {
+  if (shouldPropertyAutocomplete(value)) {
+    // +1 to account for the `.` or `[`
+    return getParentExpression(value).length + 1;
+  } else {
+    return 0;
+  }
+}
+
+export function insertAutocompleteMatch(value: string, match: string) {
+  const parentExpression = getParentExpression(value);
+
+  if (shouldPropertyAutocomplete(value)) {
+    if (isBracketNotation(value)) {
+      return `${parentExpression}[${match}]`;
+    } else {
+      return `${parentExpression}.${match}`;
+    }
+  } else {
+    // For variable autocomplete.
+    return match;
+  }
+}
+
+export function getAutocompleteMatches(input: string, frameScope: FrameScope) {
+  if (!frameScope?.scope) {
+    return [];
+  }
+
+  const bindingNames = getBindingNames(frameScope);
+
+  if (shouldPropertyAutocomplete(input)) {
+    const computedProperty = isBracketNotation(input);
+    const lastToken = getLastToken(input);
+    const parentToken = getParentExpression(input);
+
+    const binding = getBinding(parentToken, frameScope);
+
+    if (!binding) {
+      return [];
+    }
+
+    const object = binding.value._object;
+
+    if (!object) {
+      return [];
+    }
+
+    const properties = getPropertiesForObject(object);
+
+    const filteredProperties = fuzzyFilter(properties, getMatchString(lastToken));
+    return filteredProperties.map(p => (computedProperty ? `"${p}"` : p));
+  } else if (shouldVariableAutocomplete(input)) {
+    const variableNames = [...bindingNames, ...getGlobalVariables(frameScope.scope)];
```

Comment:
And honestly feels like it could just be `getBindingNames(frameScope)` here. No need for the local.